sklearn/feature_selection/tests/test_from_model.py

```diff
@@ -40,6 +41,130 @@ def test_input_estimator_unchanged():
+        transformer = SelectFromModel(estimator=est,
+                                      max_features=valid_max_n_feature,
+                                      threshold=-np.inf)
+        X_new = transformer.fit_transform(X, y)
+        if valid_max_n_feature == 'all':
+            valid_max_n_feature = max_features
+        assert_equal(X_new.shape[1], valid_max_n_feature)
+
+
+class FixedImportanceEstimator(BaseEstimator):
+    def __init__(self, importances):
+        self.importances = importances
+
+    def fit(self, X, y=None):
+        self.feature_importances_ = np.array(self.importances)
+
+
+def check_max_features(est, X, y):
+    X = X.copy()
+    max_features = X.shape[1]
+
+    check_valid_max_features(est, X, y)
+    check_invalid_max_features(est, X, y)
+
+    transformer1 = SelectFromModel(estimator=est, max_features='all',
+                                   threshold=-np.inf)
+    transformer2 = SelectFromModel(estimator=est,
+                                   max_features=max_features,
+                                   threshold=-np.inf)
+    X_new1 = transformer1.fit_transform(X, y)
+    X_new2 = transformer2.fit_transform(X, y)
+    assert_array_equal(X_new1, X_new2)
+
+    # Test max_features against actual model.
+    transformer1 = SelectFromModel(estimator=Lasso(alpha=0.025,
+                                   random_state=42))
+    X_new1 = transformer1.fit_transform(X, y)
+    scores1 = np.abs(transformer1.estimator_.coef_)
+    candidate_indices1 = np.argsort(-scores1, kind='mergesort')
+
+    for n_features in range(1, X_new1.shape[1] + 1):
+        transformer2 = SelectFromModel(estimator=Lasso(alpha=0.025,
+                                       random_state=42),
+                                       max_features=n_features,
+                                       threshold=-np.inf)
+        X_new2 = transformer2.fit_transform(X, y)
+        scores2 = np.abs(transformer2.estimator_.coef_)
+        candidate_indices2 = np.argsort(-scores2, kind='mergesort')
+        assert_array_equal(X[:, candidate_indices1[:n_features]],
+                           X[:, candidate_indices2[:n_features]])
+    assert_array_equal(transformer1.estimator_.coef_,
+                       transformer2.estimator_.coef_)
+
+    # Test if max_features can break tie among feature importance
+
+    feature_importances = np.array([4, 4, 4, 4, 3, 3, 3, 2, 2, 1])
+    for n_features in range(1, max_features + 1):
+        transformer = SelectFromModel(
+            FixedImportanceEstimator(feature_importances),
+            max_features=n_features,
+            threshold=-np.inf)
+        X_new = transformer.fit_transform(X, y)
+        selected_feature_indices = np.where(transformer._get_support_mask())[0]
+        assert_array_equal(selected_feature_indices, np.arange(n_features))
+        assert_equal(X_new.shape[1], n_features)
+
+
+def check_threshold_and_max_features(est, X, y):
```

Comment:
I'd rather this be a separate `test_threshold_and_max_features`

Reply:
I think this test currently tests threshold too much, when it is already covered above. A good set of tests should look like a proof by induction. We start by checking the basic features, and then that their combination makes sense. So I think this test would work well if we assume `threshold` and `max_features` work alone, and only confirm that their combination produces the features corresponding to the set intersection of their selections. I.e. this test should not bother with `coef_` or with `shape`.